### ISSUE TITLE
Catch more errors to avoid limbo states

### ISSUE THREAD

**Issue by davidread** [ckan/ckanext-harvest#112]
I've come a cross a bunch of ways to end up with the harvester in a 'limbo' state, so I'd like to add various exception handling and state management to avoid this better.

'limbo' is where the job state is 'running' yet something has gone wrong at it has stopped. The main problem is that you can't restart the job while it is supposedly still running.

e.g. you are writing a harvester and it has an uncaught exception in the gather, fetch or import_stage, then you enter limbo. This should be sorted better by queue.py, rather than rely on harvesters not raising exceptions.

e.g. you install a new harvester into a CKAN, restart CKAN but don't restart the queue. Running the job you get "No harvester could be found for source type x" and the job is in limbo.

**Comment by torfsen** [comment]
Exceptions from a plugin's `gather_stage`, `fetch_stage`, and `import_stage` should be caught and the appropriate action should be taken. For example, there is already error handling code in [`queue.fetch_and_import_stages`](https://github.com/ckan/ckanext-harvest/blob/13d5395bf01e80e3883e7b08b01b1cdcf5734720/ckanext/harvest/queue.py#L391), but it's only active when the plugin reports the error, not if it throws an exception. Therefore, when my plugin raised an exception from its `import_stage` it crashed the fetch consumer and left the job in limbo.

What's the current state regarding this issue?

**Comment by torfsen** [comment]
Also, the README for `gather_stage` says

> to abort the harvest, create a HarvestGatherError and raise an exception. Any created HarvestObjects will be deleted.

What it does not say is that this will crash the gather consumer because it re-raises the exception after doing the documented cleanup.

I don't see why so much of the error handling is left to the plugin implementation and why some of it is handled via return codes (i.e. returning `False` from `fetch_stage` and `import_stage`). From a plugin developer's point of view, here's how I'd like things to work:

If any of `gather_stage`, `fetch_stage` or `import_stage` raise an exception then the corresponding consumer catches it, logs it, makes sure the job and its objects are in a proper state, and then continues working. If no exception is raised then that already indicates success, so there's no need to pass an error flag via the return value.

If I'd need to create customized `HarvestGatherError` instances or some other fancy error handling then I can still do that in my plugin -- but I won't have to, and things will still work (or better, fail) correctly.

**Comment by davidread** [comment]
I agree, the harvesting framework should be defensive against harvesting code that raises exceptions.

I think there has been reluctance to catch exceptions, because it is handy during development to let an exception explode everything. Although in production this is no good because it leaves you in limbo state, like you say.

So yes, I'd been keen for you (or someone) to do a PR that catches and saves errors, like you say. You'd need to maintain backwards-compatibility with older harvesters by dealing with the response. And I'd be keen to differentiate between unintended exceptions and ones that were intended (ie have an explanation message rather than simple 'divide by 0'.)
